File: python/pheme/src/pheme_sdk/exceptions.py
"""Typed exception hierarchy for the Pheme SDK."""

from __future__ import annotations


class PhemeApiError(Exception):
    """Raised when the Pheme API returns a non-2xx response."""

    def __init__(self, status: int, message: str, body: str | None = None) -> None:
        super().__init__(f"Pheme API error {status}: {message}")
        self.status = status
        self.message = message
        self.body = body

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}(status={self.status}, message={self.message!r})"


class PhemeAuthError(PhemeApiError):
    """Raised on 401 Unauthorized or 403 Forbidden responses."""

    def __init__(self, body: str | None = None) -> None:
        super().__init__(401, "Unauthorized — check your API key or JWT token", body)


class PhemeNotFoundError(PhemeApiError):
    """Raised when the requested resource does not exist (404)."""

    def __init__(self, resource: str | None = None, body: str | None = None) -> None:
        msg = f"Not found: {resource}" if resource else "Resource not found"
        super().__init__(404, msg, body)


class PhemeRateLimitError(PhemeApiError):
    """Raised on 429 Too Many Requests. Includes *retry_after* seconds."""

    def __init__(self, retry_after: float = 1.0, body: str | None = None) -> None:
        super().__init__(429, "Rate limit exceeded", body)
        self.retry_after = retry_after

File: csharp/pheme/src/Pheme.Sdk/Exceptions.cs
// Copyright 2026 Digital Forge Studios Inc.
// SPDX-License-Identifier: MIT

namespace Pheme.Sdk;

/// <summary>Base exception for all Pheme API errors.</summary>
public class PhemeApiException : Exception
{
    /// <summary>HTTP status code returned by the server.</summary>
    public int StatusCode { get; }

    /// <summary>Raw response body, if available.</summary>
    public string? ResponseBody { get; }

    public PhemeApiException(int statusCode, string message, string? responseBody = null)
        : base(message)
    {
        StatusCode = statusCode;
        ResponseBody = responseBody;
    }
}

/// <summary>Thrown when the server returns HTTP 401 Unauthorized.</summary>
public sealed class PhemeAuthException : PhemeApiException
{
    public PhemeAuthException(string? responseBody = null)
        : base(401, "Authentication failed — check your API key or JWT.", responseBody) { }
}

/// <summary>Thrown when the server returns HTTP 403 Forbidden.</summary>
public sealed class PhemeForbiddenException : PhemeApiException
{
    public PhemeForbiddenException(string? responseBody = null)
        : base(403, "Access denied.", responseBody) { }
}

/// <summary>Thrown when the server returns HTTP 404 Not Found.</summary>
public sealed class PhemeNotFoundException : PhemeApiException
{
    public PhemeNotFoundException(string resource, string? responseBody = null)
        : base(404, $"Resource not found: {resource}", responseBody) { }
}

/// <summary>Thrown when the server returns HTTP 429 Too Many Requests.</summary>
public sealed class PhemeRateLimitException : PhemeApiException
{
    /// <summary>Seconds to wait before retrying, if the server supplied a Retry-After header.</summary>
    public int? RetryAfterSeconds { get; }

    public PhemeRateLimitException(int? retryAfterSeconds = null, string? responseBody = null)
        : base(429, $"Rate limit exceeded. Retry after {retryAfterSeconds?.ToString() ?? "unknown"} seconds.", responseBody)
    {
        RetryAfterSeconds = retryAfterSeconds;
    }
}

/// <summary>Thrown for unexpected server errors (5xx).</summary>
public sealed class PhemeServerException : PhemeApiException
{
    public PhemeServerException(int statusCode, string? responseBody = null)
        : base(statusCode, $"Server error: HTTP {statusCode}", responseBody) { }
}

/// <summary>Thrown when the HTTP request cannot be completed (network failure, timeout, etc.).</summary>
public sealed class PhemeNetworkException : Exception
{
    public PhemeNetworkException(string message, Exception? inner = null)
        : base(message, inner) { }
}
